<h1 align=center><font size = 5> Build a Regression Model with Keras with Data Normalization and Increased number of Epochs</font></h1>

# <h1 align=left><font size = 3>C. Increase the number of epochs</font></h1>


Use the Keras library to build a neural network with the following:
- One hidden layer of 10 nodes, and a ReLU activation function
- Use the **adam** optimizer and the **mean squared error** as the loss function.


Let's start by importing the pandas and the Numpy libraries.

In [1]:
import pandas as pd
import numpy as np

Let's download the data and read it into a <em>pandas</em> dataframe.



In [2]:
concrete_data = pd.read_csv(' https://cocl.us/concrete_data')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [3]:
concrete_data.shape

(1030, 9)

Let's check the dataset for any missing values.


In [4]:
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [5]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

The data looks very clean and is ready to be used to build our model.


**Split data into predictors and target and randomly spliting the data into a training and test sets by holding 30% of the data for testing.**

In [6]:
from sklearn.model_selection import train_test_split

In [7]:
concrete_data_columns = concrete_data.columns

predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

X_train, X_test, y_train, y_test = train_test_split( predictors, target, test_size=0.3, random_state=30)

In [8]:
n_cols = predictors.shape[1] # number of predictors

To normalize the data by substracting the mean and dividing by the standard deviation.

In [9]:
X_train_norm = (X_train - X_train.mean()) / X_train.std()
X_train_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
590 -0.419281           -0.852944 -0.859625  0.207241         -1.077712   
197 -0.839210           -0.852944  0.723351 -0.720695          0.194690   
184 -0.571191           -0.852944  0.663497  0.373439         -0.314271   
57  -0.801474            0.688746 -0.859625  0.498087         -1.077712   
200 -0.877913           -0.852944  1.115551 -0.882276          0.245586   

     Coarse Aggregate  Fine Aggregate       Age  
590          1.890821        0.202965 -0.623796  
197          0.424266        1.675144  0.219283  
184         -0.088507        1.225662 -0.692619  
57           0.058932        0.660022 -0.692619  
200          1.515049        0.388566 -0.503356

In [10]:
X_test_norm = (X_test - X_test.mean()) / X_test.std()
X_test_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
3    0.484070            0.768370 -0.816436  2.210307         -0.952639   
517 -0.731903           -0.737481  1.344010  1.143961         -0.676270   
107  0.402074            2.374992 -0.816436  0.067921          0.721832   
659 -1.604981            0.996252 -0.816436  1.022785         -0.952639   
381  0.321009            0.705388 -0.816436 -1.812725          0.006524   

     Coarse Aggregate  Fine Aggregate       Age  
3           -0.481133       -2.186062  4.273190  
517         -0.356708        0.322187 -0.603113  
107         -0.347999       -1.387542 -0.603113  
659         -0.403990        0.903809  0.527426  
381          1.982477       -0.356373 -0.317073

We can now import keras and its classes Sequential and Dense

In [11]:
import keras

In [12]:
from keras.models import Sequential
from keras.layers import Dense

We can now build our neural network :

In [13]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,))) # One hidden layer of 10 nodes, and a ReLU activation function
    model.add(Dense(1))

    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

Train the model on the training data using **100 epochs** while randomly spliting the data into a training and test sets by holding 30% of the data for testing.

In [14]:
# build the model
model = regression_model()

In [ ]:
# fit the model
model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=100, verbose=2)

Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength. You can use the mean_squared_error function from Scikit-learn.

In [16]:
from sklearn.metrics import mean_squared_error

# Initialize an empty list and a counter
MSE_list = []
counter = 1

# Fill the list with elements using a while loop
while counter <= 50:
    model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=100, verbose=2)
    y_pred = model.predict(X_test)
    MSE_list.append(mean_squared_error(y_test, y_pred))
    counter += 1

print(MSE_list)

Le flux de sortie a été tronqué et ne contient que les 5000 dernières lignes.
Epoch 14/100
23/23 - 0s - loss: 45.2741 - val_loss: 52.6986 - 82ms/epoch - 4ms/step
Epoch 15/100
23/23 - 0s - loss: 47.4631 - val_loss: 52.4809 - 86ms/epoch - 4ms/step
Epoch 16/100
23/23 - 0s - loss: 50.3579 - val_loss: 60.4841 - 103ms/epoch - 4ms/step
Epoch 17/100
23/23 - 0s - loss: 45.9805 - val_loss: 54.4978 - 110ms/epoch - 5ms/step
Epoch 18/100
23/23 - 0s - loss: 44.6908 - val_loss: 57.4169 - 94ms/epoch - 4ms/step
Epoch 19/100
23/23 - 0s - loss: 45.6526 - val_loss: 52.7631 - 104ms/epoch - 5ms/step
Epoch 20/100
23/23 - 0s - loss: 44.8613 - val_loss: 53.7926 - 108ms/epoch - 5ms/step
Epoch 21/100
23/23 - 0s - loss: 44.3655 - val_loss: 54.3770 - 97ms/epoch - 4ms/step
Epoch 22/100
23/23 - 0s - loss: 45.1320 - val_loss: 52.4862 - 108ms/epoch - 5ms/step
Epoch 23/100
23/23 - 0s - loss: 49.2663 - val_loss: 56.4181 - 94ms/epoch - 4ms/step
Epoch 24/100
23/23 - 0s - loss: 49.2421 - val_loss: 57.5437 - 92ms/epoch - 4m

In [17]:
import statistics
mean = statistics.mean(MSE_list)
standard_deviation = statistics.stdev(MSE_list)

print("Mean after Normalization and with 100 epochs:", mean)
print("Standard Deviation after Normalization and with 100 epochs:", standard_deviation)

Mean after Normalization and with 100 epochs: 67.28571954024014
Standard Deviation after Normalization and with 100 epochs: 73.69759301609396


**Part A:**

Mean: 61.02331363622501

Standard Deviation: 14.615134162091586

**Part B:**

Mean after Normalization: 58.20324839422981

Standard Deviation after Normalization: 10.494701801911273

**Part C:**

Mean after Normalization and with 100 epochs: 67.28571954024014

Standard Deviation after Normalization and with 100 epochs: 73.69759301609396

**Conclusion**

- Normalizing the input data allowed us to have more precise results
- Increasing the number of epochs 50 --> 100 made us reach lower loss values but still shows inconsistent MSE values
- Increasing the number of epochs isn't always the solution for a better accuracy and model performance and it lead in this exemple in overftiing: The model seems to be memorising the data instead of learning it.